### DISCUSSION TITLE
RFC: New Tasklist User Interface

### DISCUSSION THREAD

**Discussion by jaap-karssenberg** [zim-desktop-wiki/zim-desktop-wiki#1863] [Ideas]
Dear all,

In the zim development branch we are testing a new user interface for the tasklist. I would like to get feedback from users trying it out before merging it in the main branch and releasing it. Please try it out and let me know your thoughts how to improve further.

Key changes:
- New interface for selecting "lists" "tags" and "pages"
- Optional "GTD" flavor lists in the selection pane for "next" / "inbox" / "waiting"
- New menu in the dialog to allow also including closed tasks in the overview
- Improved parsing of task labels
- Allow inheritance of all modifiers (tags, due date, start date etc.) from a label starting a list of tasks
- Tasks appearing in a weekly journal page can have automatic due or start date per day instead of per week

For more explanation, please see the manual page [here](https://raw.githubusercontent.com/zim-desktop-wiki/zim-desktop-wiki/develop/data/manual/Plugins/Task_List.txt) (sorry for the raw formatting)

**Screenshot side pane**
![Screenshot side pane](https://user-images.githubusercontent.com/1248286/149397079-cd8237c0-9df7-41c1-a02f-ac4690b3d2ea.png)

**Screenshot of tasklist dialog**
![Screenshot of tasklist dialog](https://user-images.githubusercontent.com/1248286/149397101-d65e097b-73d4-440c-bf2a-233b742fa325.png)

Development branch source: [github](https://github.com/zim-desktop-wiki/zim-desktop-wiki/tree/develop)
Development branch PPA: [launchpad](https://launchpad.net/~jaap.karssenberg/+archive/ubuntu/zim-snapshots)

Unfortunately there is no development branch windows installer or Mac OS application at this time

Kind regards,

Jaap

**Comment by reagle** [comment]
What's a trans-migrated task?

**Comment by jaap-karssenberg** [comment]
The "migrated" and "trans-migrated" tasks are used in the Bullet Journal
methodology - there is a separate manual page on that topic if you are
interested (and a lot of generic internet pages).

**Reply by reagle** [reply]
For other, I found this: https://bulletjournal.com/pages/migration

**Comment by treborxy** [comment]
Hi Jaap, thanks again for your work - at a first view it looks like a very useful improvement!
One thing about developement - branches: I suppose its my fault by misconfiguring something - but when I installed the testing-version at my production-machine, it started without any hesitating using my zim production-data. It looks like nothing bad happened - but how could I prevent future dev-versions to do so?
Best wishes, Robert

**Reply by jaap-karssenberg** [reply]
Hi Robert,

One way is to create s test notebook and when testing the dev version provide it as command line argument so Zim opens the test notebook instead of the default notebook 

The other way is to provide a full alternative environment using the XDG environment variables. The manual had more detail about these.

However in general I would not expect the development version to corrupt data. I use the development version always for my own real job. Of course there is always a risk for bugs, so best advice I can give is to rename the version control plugin and ensure days is committed before switching versions. That easy you can always roll back.

Regards,

Jaap

**Comment by introt** [comment]
Hi Jaap, and congrats on the revamped tasklist plugin, it's really great! Automatic per-day due dates on week pages have really elevated my day planning game, and the new side pane makes it really easy to quickly find all tasks from a given page hierarchy.

**Bugs**

I'm seeing greyed-out tasks in the Active list. They inherit one of my `non_actionable_tags` from their parent. Given the option to set these tags is missing from the UI, seems they are to be migrated to `waiting_labels`?

**Usability**

Resizing the Task window makes its sidebar grow too wide; it should behave similar to the panels and keep its (user set) max width.

Currently, the pages are listed alphabetically in a flat list. Could there be an optional tree view? It's somewhat confusing to see a top-level page with a task count in tens sandwiched between its sub-sub-pages with a handful of tasks each. Single-click to select and double click to expand?

Related to handling large amounts of tasks, having those numbers reflect the number of tasks matching the filter (as they do when selecting tags) would make sense in my opinion.

Regarding the task list itself, I'm having problems with urgent sub tasks getting buried due to the trees in `All tasks` view being sorted by the urgency of their top-level task. Making list flatness implicit via the removal of the `Show tasks as flat list` option didn't help here, though it wasn't a complete solution either as you lose the context of the sub task.

Other improvements that come to mind would be alternating the list background colors for readability (& perhaps to convey extra information, such as highlighting pages with tasks due soon) and being able to choose as well as reorder the columns by using a context menu and dragging. Having tasks with links to pages appear under those pages' lists (or alternatively, a week view) would be awesome for week planning.


Looking forward to the next release!

**Reply by jaap-karssenberg** [reply]
Could you open "[develop]" tickets for (a) the handling of non-actionable-tasks and (b) the behavior of the side pane when resizing ? Would like to fix those before releasing.

With respect to the "flat list" this is what the "Active" list is now. "Active", "Inbox" & "Next actions" are flat always. My thinking was that the toggle in a context menu was to much hidden. Therefore I made it a separate list type in the selection view. Does this work for you, or do you miss something additionally ?

**Reply by jaap-karssenberg** [reply]
Side pane behavior is fixed.

On the topic of the "Waiting" category - should tasks that are not yet started also be expected here, are is that a separate category ?

**Reply by introt** [reply]
I've opened #1917, sorry for the delay!

---

> With respect to the "flat list" this is what the "Active" list is now. "Active", "Inbox" & "Next actions" are flat always. My thinking was that the toggle in a context menu was to much hidden. Therefore I made it a separate list type in the selection view. Does this work for you, or do you miss something additionally ?

My problem is the display order of the following tasks:

        [ ] complete all following before exam < 2021-05-01
            [ ] subtask to be completed way before exam < 2021-03-01

        [ ] another task < 2021-04-01

Currently, "another task" appears before "subtask to be completed way before exam" when sorting by Due, because sorting happens according to the due dates of the top level tasks. I can add more examples later if needed.

**Reply by introt** [reply]
> On the topic of the "Waiting" category - should tasks that are not yet started also be expected here, are is that a separate category ?

Sorry, not familiar enough with GTD to really comment. Only came across "Waiting For [other people to do their part]" on my cursory searches, which would point to "no". I'm used to using various non-actionable tags (in combination with the journal and tag pane) for all sorts of blocked items, including "Someday/Maybe" stuff, to keep them out of sight and mind.

**Comment by ZtzEAcRsND** [comment]
* General
	* I looked at the develop branch in xubuntu 21.04. 
	* This new task filter is amazing. Apparently I use very little of the current task plug-in capabilities (after noting @introt's comments too). The new format will help me clean up disparate open tasks and tags spread throughout my journal and regular zim pages.  I'll keep using this as my main zim. 
	* This layout and functionality concept should used for other core plug-ins. 
	* With the new task filter options (e.g., waiting) a kanban type display would be possible. 
	* The ability as separate window to be full screen is great as it allows more context to be displayed as desired.
	* Hiding non-relevant filters after selecting one or more filters, similar to current tag plug-side bar, is great. Current implementation might have delay between selection and display.  
	* Maintaining the current task plug-in panel view as default with the task window breakout is a good thing. Forcing a window breakout could interrupt workflow. 
* Suggestions
	* Only **allow single instance of task list window**. Multiple task list windows are possible which may be confusing. Maybe make 'single' or 'multiple' task filter windows optional.
	* **Task filter selection** (balloon window), when selected and viewed as embedded in a side panel, displays downward, not to the side, covering the task list. This might not be a problem since we are selecting the filters and toggle the task filter off. The task filter selection does not cover the task list in a top or bottom panel. Let me know if you'd like me to explore other variables. I can't get the same side-by-side display you've shown.
	* The **Task Status filter options (open, closed, canceled, migrated, trans) **displayed under the hamburger menu (gear icon on my xfce theme). These seem to be filter options that could be better placed in the filter list placed under "All Tasks" and perhaps documented as 'bullet list' filter options as opposed to GTD filter options. The "Clear Selections" would be available to the Task Status options if moved to the main Task Filter window.
	* If a **hamburger menu list is used, consider using it for task list options** such as select/de-select filter groups in the task filter selection window, instead of additional filters such as Task status.  Maybe the GTD Task option could go here (too?).
	* **If this is the first hamburger menu in zim**, documentation for what to expect and how to access the hamburger menu might be helpful. e.g., the shortcut key to access the hamburger menu.
	* **Make the filter list collapsible (treeview?)** such that the filter items can be collapsed under the filter type heading (e.g., tag, page). Similar to the Page Index and Tags pane.  Navigation (collapse, expand, next/previous filter, select) could be similar to the page index. However, a collapsible filter list doesn't display well during filter selection causing other filters options to get removed from the list if unrelated.  And, however, a collapsible list may be unweldy for many levels of sub pages.
	* **Count tasks per page without counting tasks on sub-pages.** e.g., I have over 1,000 tasks in a parent folder with many sub-folders; maybe because I've redefined a different Home Page.
	* **Add a filter option for task priority.** The task priority is already a sortable column.
	* **Add filter page option 'current page'** (or as an option, the last multiple current pages) which could be the page(s) from the Page Index or Journal. 
	* Consider using a **different icon for open tasks** (maybe an image version of []), rather than a bullet.
* Not clear or obvious
	* **Keyboard navigation** in the task filter pane doesn't seem consistent. e.g., tab, up/down/left/right arrow, space bar (selection?).  Some inconsistency may be performance delay related and my impatience. 
	* **keyboard navigation between task window panes** isn't obvious to me e.g., <ctrl>tab or arrow keys. Maybe this is cause by the delay?
	* There is a **noticeable delay** (multiple seconds)to display the task window breakout when selected. As a system/zim performance reference, most other zim operations on this system are nearly instantaneous.
	* There is a **noticeable delay** (multiple seconds) to display selection after selecting any filter option including (especially) the filter options under the hamburger menu.
	* I'm not clear on how to use GTD labels like 'active' and 'project'--this is probably my problem and need to read the documentation more closely. 

Thanks. I hope this is helpful.

**Reply by jaap-karssenberg** [reply]
Thanks for the feedback. I'll keep your suggestions in mind. However some of the behavior (like placement of the "balloon") is Gtk toolkit behavior, not easily modified from application code.

Few specific comments:
* The "open tasklist" item in the window header bar or toolbar opens a single instance. However the side pane "pop out" opens new ones because is copies the selection of the pane and I didn't want to overwrite the selection you may have already open in the dialog.
* The page names are used as "tags" and therefore explicitly child pages are counted as well. Modifying this will make it just regular page selection which is less interesting - but should be possible as well (my plan is to add better search capabilities in the filter, to allow filtering by attributes - similar as the key-value syntax in the search window)
* Agree the icons for open / closed / etc. are not optimum - couldn't find a better fit on short notice. Tried using the checkboxes, but the problem is that that suggest they are clickable, so it might confuse users. I'm open for suggestions what to use instead.

Kind regards,

Jaap

**Comment by victorfeight** [comment]
Fantastic additions, I'll be testing them out in the coming days. I'm eagerly awaiting a save state between task usage .functionality

**Comment by releve1403** [comment]
Hi Jaap, Very good upgrade to the task plugin! 

I like very much the possibility to see closed tasks (had to do a content search on Zim text files lo locate those). 

Only thing missing is the total number of tasks with was present in the preceding versions. Also, I may be wrong but, with the old versions, we could write directly to filter tasks (it was in filter mode by default, I think).

And, for whatever reason, printing tasks list produce an error in my default browser (Firefox).

I then tried opening the file named "print-to-browser.html" in /tmp and I got the same error (ERR_FILE_NOT_FOUND) with :
* Firefox
* Chromium

But I can open the file without any problem with the following browsers :
* Google Chrome
* Opera
* GNOME web browser (epiphany)

Very good improvement.

**Reply by jaap-karssenberg** [reply]
Sounds like a bug that is not related directly to this new interface - does it work correctly in the released version ?  If so, please open a ticket for this.

**Reply by releve1403** [reply]
Hi, 
In fact, the behavior is the same with the master version. So it's not related to Zim but more on the way browsers handle local files like "/tmp/zim-bvnbpc65/print-to-browser.html". I will search on this side.
So there is no need to open a ticket for this.
Thank you and cheers.
--Renaud

**Reply by releve1403** [reply]
Hi again,
Also, after checking the master version :
1. Total number of tasks is present
2. Text filter is active by default
(see joined file)
How to ask for these? Should I open an issue? I'm not familiar with the process on Github.
Thanks
--Renaud
![Capture de 2022-02-13 19:27:23](https://user-images.githubusercontent.com/98767445/153790516-f9d2fe2a-4ebc-4ca3-954b-b7efcbb9a81a.png)

**Reply by introt** [reply]
Hi @releve1403! You can open `[develop]` tickets for these [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/new?assignees=&labels=bug%2C+needs+triage&template=bug-report-for-development-branch-or-snapshot-ppa.md&title=%5Bdevelop%5D) (if the link doesn't work, just [open a new ticket](https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/new/choose) and "choose Bug report for development branch or snapshot PPA". It's preferred to open an issue per item, but feel free to share the screenshot between the two.

**Comment by bruvellu** [comment]
Hey Jaap,

Great improvements to the task list plugin. The daily due dates on the week page help me a lot with task sorting. Thanks for working out the implementation. I've been testing the other features and I have one main suggestion:

A way to automatically filter tasks based on the current page (including child pages).

Here's my typical workflow where this feature would be handy:

It's a new day. I first check my planned tasks on my journal's weekly page. Then, I look to the tasklist sidebar to check tasks further in the future or any pending higher-priority tasks to act on. With this I get the global overview. If it's a normal day, I go back to the week page and pick the task to work on. I often have `[ ] Do X for Project Y` with a link to the project page. I click and open the project page. That's where I keep the project's main tasks; child pages contain specific tasks. Now, I want to focus on the project's tasks, but the sidebar is still showing the notebook-wide tasks.

The sidebar page filtering already solves this, but it's a manual process. It takes time to click the dropdown menu and scroll through a long list before finding the page name. If I need to switch between projects (pages) more than a few times, it soon becomes unpractical. I also noticed that I rarely change the page filter to something else other than the current page. Automatic filtering based on the current page would make the process smoother.

An idea is to add a "Current Page" as the first item on the page filtering list. It could use the same backend of the page filtering, but get and pass the current page name to the selected pages.

Let me know what do you think!

Cheers,
Bruno

**Reply by jaap-karssenberg** [reply]
Adding more defaults filters is a good idea. 
Currently you can also do this by typing a page name in the filter should work since the filter also checks against the "page" column.

**Comment by ZtzEAcRsND** [comment]
@jaap-karssenberg: On the task page selection pane, alike page names, but different paths are shown in the same task list. e.g., selecting a date (using daily  journal pages (01, 02, etc)) would display all 01 day pages for different months. I can't find an easy way to isolate the tasks for a specific day/month.  A similar case: Path:Names like deb:package vs arch:package.   Currently, selecting multiple pages in the task page selection pane displays all tasks from all selected pages. 

If there's not a way to isolate a set of tasks a couple possible fixes: (1) Prepending some of the path (e.g., 2022-02 vs 2021-02) as the pathbar button naming attempts to do OR (2) providing an option to change the task page selection logic to AND instead of OR (similar to the tag plug-in selection).   would help (1) identify different pages and (2) could be used to group sort pages in the task page selection pane.  The path bar button naming attempts to do this. 

I understand option (2) is different than the current implementation and might break the project task collection capability.  

![Screenshot_2022-02-10_08-46-08](https://user-images.githubusercontent.com/91022901/153420591-25b4f8a9-65b0-4601-bcae-a6c614615875.png)

**Reply by introt** [reply]
> I can't find an easy way to isolate the tasks for a specific day/month.

There's a workaround: select the month/day number and then sort using the page column (that one has the full paths).

> as the pathbar button naming attempts to do

That one uses [shortest_unique_names](https://introt.github.io/zim-desktop-wiki/docs/api/zim.notebook.page.html#shortest_unique_names), which could get a little silly if eg. there are tasks on both day and month pages: unique 12 (month) and unique 11 (day) look alike. Also, it would prevent using page names for categorization (eg. listing all Next tasks under multiple projects (Project1:Next and Project2:Next); not a great example, but I hope you get the gist).

I'd go for a third option, excluding auto-generated (eg. not custom subpages) Journal pages from the Page list, and adding a separate Journal list with full-er paths, or even a calendar widget.

**Reply by ZtzEAcRsND** [reply]
Thanks for the response. I appreciate that this "isolate task" problem may be unique to me.  

Your "page names for categorization" example is a good counter-example for my suggestion.  That approach must work for most people.  I don't  organize projects or group tasks that way.

I'd make the Journal plug-in modification  modification a really low priority. I only activate the Journal plug-in to get the ALT-D shortcut and set the option to 'monthly'. The rest of Journal could be hidden or go away. As you indicate the non-problem is not unique to tasks in the journal pages, but probably where most of the non-unique page names are. 

I did see the 'sort column' capability.  A fourth option would be to drop to the terminal and use 'find'.

**Comment by rolfkleef** [comment]
If I view only open tasks, and select a tag, I can quickly jump to an open task with that tag.

However, when I check that task as completed, the task list switches back to "no tag selected", with all open tasks. Would be nice to keep the tag selected, I haven't found a way for that yet.

**Comment by dehesselle** [comment]
I've built a macOS app from current development branch for testing purposes: https://gitlab.com/dehesselle/zim_macos/-/jobs/2202268681/artifacts/raw/Zim.dmg
_(unsigned; "about" dialog will say "Zim 0.74.3 (57)")_

I've noticed that the task list window presents itself in native GTK 3 CSD style, I'm guessing it doesn't know that I've disabled that for Zim itself.
![zim_tasklist](https://user-images.githubusercontent.com/11375494/158274819-9c491466-e69c-41b9-8a36-bd9dbbecfb36.jpg)

**Comment by jaap-karssenberg** [comment]
Rene, can you open an issue ticket for this ?  Clearly I need to do some
further re-factoring on the mechanism for setting this option.

-- Jaap

**Reply by dehesselle** [reply]
Done! https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1983

**Comment by Elliria** [comment]
Hey there. I see that this is already available and I'm loving using it. Thank you so much for creating it.I hope it's not too late to join in on the feedback and ask a question. Is there a way that we can use CSS or some other method to get the other columns to inherit the red, orange, and yellow colors that are automatically applied to the Date column so that all columns receive matching automatic colors? If not, I'd like to make that a wish.
